### PR TITLE
adding get actor reminder API

### PR DESCRIPTION
**Pull request by shivamkm07**
# Description

It adds the get actor reminder API in dotnet-sdk

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #772 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: src/Dapr.Actors/Runtime/ReminderInfo.cs [hunk 1]
```diff
@@ -55,28 +55,29 @@ internal static async Task<ReminderInfo> DeserializeAsync(Stream stream)
...
-            if (json.TryGetProperty("data", out var dataProperty) && dataProperty.ValueKind != JsonValueKind.Null)
-            {
-                data = dataProperty.GetBytesFromBase64();
-            }
-
-            if (json.TryGetProperty("ttl", out var ttlProperty))
-            {
-                var ttlString = ttlProperty.GetString();
-                ttl = ConverterUtils.ConvertTimeSpanFromDaprFormat(ttlString);
+            if(json.ValueKind != JsonValueKind.Null){
```
Comment (philliphoff) [inline src/Dapr.Actors/Runtime/ReminderInfo.cs:58]: Would it be more accurate to say?: `if (json.ValueKind == JsonValueKind.Object)`

Comment (philliphoff) [inline src/Dapr.Actors/Runtime/ReminderInfo.cs:58]: `DeserializeAsync<T>()` returns `object?`; would it make sense to also test `json` for `null`?

File: src/Dapr.Actors/Runtime/Actor.cs [hunk 2]
```diff
@@ -360,6 +360,18 @@ internal async Task<IActorReminder> RegisterReminderAsync(ActorReminderOptions o
...
         }
 
+        /// <summary>
+        /// Gets a reminder previously registered using <see cref="Dapr.Actors.Runtime.Actor.RegisterReminderAsync(ActorReminderOptions)" />.
+        /// </summary>
+        /// <param name="reminderName">The name of the reminder to get.</param>
+        /// <returns>
+        /// Returns a task that represents the asynchronous get operation. The result of the task contains the reminder if it exists, otherwise null.
+        /// </returns>
+        protected async Task<IActorReminder> GetReminderAsync(string reminderName)
```
Comment (philliphoff) [inline src/Dapr.Actors/Runtime/Actor.cs:370]: This method states it returns `null` if no timer exists, but the underlying call to the `DaprHttpInteractor` appears to throw if the HTTP request does not return a successful response (and the docs say the Actor Get Reminder API returns only `200` and `500` responses.

File: src/Dapr.Actors/Runtime/ActorTimerManager.cs [hunk 3]
```diff
@@ -27,6 +27,13 @@ public abstract class ActorTimerManager
         /// <returns>A task which will complete when the operation completes.</returns>
         public abstract Task RegisterReminderAsync(ActorReminder reminder);
 
+        /// <summary>
+        /// Gets a reminder previously registered using
+        /// </summary>
+        /// <param name="reminder">The <see cref="ActorReminderToken" /> to unregister.</param>
+        /// <returns>A task which will complete when the operation completes.</returns>
+        public abstract Task<IActorReminder> GetReminderAsync(ActorReminderToken reminder);
```
Comment (philliphoff) [inline src/Dapr.Actors/Runtime/ActorTimerManager.cs:35]: The underlying `IDaprInteractor` allows for cancellation, but the higher-level types don't seem to make use of it.  Should we start doing so for new APIs?  It'd likely be breaking to do so as a whole, but one has to start somewhere.

File: src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs [hunk 4]
```diff
@@ -77,5 +91,24 @@ private async ValueTask<string> SerializeReminderAsync(ActorReminder reminder)
...
+
+        private async ValueTask<ActorReminder> DeserializeReminderAsync(Stream stream, ActorReminderToken token)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+            var streamContents = new byte[stream.Length];
+            stream.Read(streamContents);
+            if(Encoding.ASCII.GetString(streamContents) == "null")
```
Comment (philliphoff) [inline src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs:103]: Would the JSON deserializer not catch this case?  In particular, can we assume `ASCII` encoding vs. `UTF8`?

File: src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs [hunk 5]
```diff
@@ -77,5 +91,24 @@ private async ValueTask<string> SerializeReminderAsync(ActorReminder reminder)
...
+            {
+                return null;
+            }
+            stream.Seek(0, SeekOrigin.Begin);
```
Comment (philliphoff) [inline src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs:107]: There are no guarantees that the `Stream` provided supports seeking; if we really need to do this, the capability should be verified and, if not, a copy of the stream (e.g. a `MemoryStream`) should be made.  If we were to know the JSON deserializer handles `null`, then it seems we would have no need for this.

File: src/Dapr.Actors/Runtime/ReminderInfo.cs (hunk 1 above)
Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/ReminderInfo.cs:58]: Here we have been using `DeserializeAsync<JsonElement>` which returns `JsonElement` type, which is not a reference type and so can't be compared to null. It says this error on trying: `Operator '!=' cannot be applied to operands of type 'JsonElement' and '<null>'`.

File: src/Dapr.Actors/Runtime/Actor.cs (hunk 2 above)
Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/Actor.cs:370]: In case the reminder doesn't exist, the HTTP call TO Dapr returns a successful status code with a response {"null"}. This is being handled in the `DeserializeReminderAsync` function in ActorTimerManager(https://github.com/dapr/dotnet-sdk/pull/1103/files#diff-89d1bcb06652638fdfeaa1b8146ada6ea2d8fe4ff18216de0928734712473cf4R103), in which we actually return a null object. Also the same scenario is added in E2E test as well, testing both cases when the reminder doesn't exist and when reminder exists.
The E2E test case for no reminder can be found here: https://github.com/dapr/dotnet-sdk/pull/1103/files#diff-6c279ec059b177b7a7e54e34eda712f299033fb2321fac66836f93a78c76b5d5R62

File: src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs (hunk 4 above)
Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs:103]: No the Json Deserializer actually checks for the values in the stream and if they are not there, it assigns a default value for some(e.g. TimeSpan,Period). Also note that, in case a reminder is not there, `stream` is not null, instead it's actually has a string "null", which is what Dapr returns when the reminder is not there. So this is required for handling the case when there is no reminder.

Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs]: Handled the case in JsonDeserializer only. That's a much better and cleaner way. Thanks!

File: src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs (hunk 5 above)
Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/DefaultActorTimerManager.cs]: Removed the Seeking code, and handled the case in JsonSerializer itself.

File: src/Dapr.Actors/Runtime/Actor.cs (hunk 2 above)
Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/Actor.cs:370]: If the reminder doesn't exist, Dapr returns a suceessful status code with {null} response. That scenario is already handled in Json Serializer and so in case a reminder doesn't exist, this function returns a null value only.

Alos, the same has been tested in added E2E test as well and it is passing as well: https://github.com/dapr/dotnet-sdk/pull/1103/files#diff-6c279ec059b177b7a7e54e34eda712f299033fb2321fac66836f93a78c76b5d5R61

File: src/Dapr.Actors/Runtime/ReminderInfo.cs (hunk 1 above)
Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/ReminderInfo.cs]: Handled null scenario differently.

Comment (shivamkm07) [reply src/Dapr.Actors/Runtime/ReminderInfo.cs]: Handled null scenario differently

**Review by philliphoff — APPROVED** [review]
It might be good to more clearly document the behavior of Dapr when retrieving actor reminders that do not exist; a `200` response with a `null` body doesn't seem intuitive to me as opposed to, say, a `404` response.

**Review by yaron2 — APPROVED** [review]

**Comment by artursouza** [comment]
Why is this feature needed? We don't have it in other SDKs.

**Comment by shivamkm07** [comment]
> Why is this feature needed? We don't have it in other SDKs.

It was an ask from customers: https://github.com/dapr/dotnet-sdk/issues/772
